compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -185,10 +222,40 @@ static bool safeToMoveGuard(TR::Block *destination, TR::TreeTop *guardCandidate,
              tt->getNode()->getInlinedSiteIndex() != TR::comp()->getInlinedCallSite(guardCandidate->getNode()->getInlinedSiteIndex())._byteCodeInfo.getCallerIndex())))
             return false;
 
+         // Collect privarg store symrefs
+         if (tt->getNode()->chkIsPrivatizedInlinerArg())
+            privArgSymRefs.set(tt->getNode()->getSymbolReference()->getReferenceNumber());
+
          if (tt->getNode()->isNopableInlineGuard()
              && tt->getNode()->getBranchDestination() != branchDest)
             return false;
          }
+
+      // Search for loads of the collected privargs in the taken side of guard1
+      // This is possible if a privarg temp was recycled for the inner call site, with a prior use as an
+      // argument for the outer call site. As the privargs for the inner call site must be evaluated before
+      // both guards, this would result in the recycled temp holding the incorrect value if the guard is ever
+      // taken. Give up if this situation is detected.
+      TR_Stack<TR::Block*> blocksToCheck(TR::comp()->trMemory(), 8, false, stackAlloc);
```

Comment:
would it be cheaper to flip this the other way - gather those used on the taken side of the guard when we pick a new first guard and have a candidate to merge and keep reusing the set for the subsequent priv args? Might be lower in compile time overall, but possibly more complex. Thoughts?